storage/src/mmr/mem.rs

```diff
@@ -0,0 +1,854 @@
+//! change and can be used as stable identifiers.
+//!
+//! The "height" of a node is 0 for a leaf, 1 for the parent of 2 leaves, and so on.
+//!
+//! The "root hash" of an MMR is the result of hashing together the size of the MMR and the hashes
+//! of every peak in decreasing order of height.
+
+use crate::mmr::{Hash, Hasher, Proof};
+
+/// Implementation of `InMemoryMMR`.
+pub struct InMemoryMMR<const N: usize, H: Hasher<N>> {
+    hasher: H,
+    // The nodes of the MMR, laid out according to a post-order traversal of the MMR trees, starting
+    // from the from tallest tree to shortest.
+    nodes: Vec<Hash<N>>,
+}
+
+/// A PeakIterator returns a (position, height) tuple for each peak in the MMR, in decreasing order
+/// of height. The iterator will only return the peaks that existed at the time of its
+/// initialization if new elements are added to the MMR between iterations.
+#[derive(Default)]
+struct PeakIterator {
+    sz: u64,       // number of nodes in the MMR at the point the iterator was initialized
+    node_pos: u64, // position of the current node
+    two_h: u64,    // 2^(height+1) of the current node
+}
+
+impl PeakIterator {
+    /// Return a new PeakIterator over the peaks of a MMR with the given number of nodes.
+    fn new(sz: u64) -> PeakIterator {
+        if sz == 0 {
+            return PeakIterator::default();
+        }
+        // Compute the position at which to start the search for peaks. This starting position will
+        // not be in the MMR unless it happens to be a single perfect binary tree, but that's OK as
+        // we will descend leftward until we find the first peak.
+        let start = u64::MAX >> sz.leading_zeros();
+        let two_h = 1 << start.trailing_ones();
+        PeakIterator {
+            sz,
+            node_pos: start - 1,
+            two_h,
+        }
+    }
+}
+
+impl Iterator for PeakIterator {
+    type Item = (u64, u32); // (peak, height)
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while self.two_h > 1 {
+            if self.node_pos < self.sz {
+                // found a peak
+                let peak_item = (self.node_pos, self.two_h.trailing_zeros() - 1);
+                // move to the right sibling
+                self.node_pos += self.two_h - 1;
+                assert!(self.node_pos >= self.sz); // sibling shouldn't be in the MMR if MMR is valid
+                return Some(peak_item);
+            }
+            // descend to the left child
+            self.two_h >>= 1;
+            self.node_pos -= self.two_h;
+        }
+        None
+    }
+}
+
+/// A PathIterator returns a (parent_pos, sibling_pos) tuple for the sibling of each node along the
+/// path from a given perfect binary tree peak to a designated leaf, not including the peak itself.
+///
+/// For example, for the tree below, the path from the peak to leaf node 3 would produce:
+/// [(6, 2), (5, 4)]
```

Comment:
nit: you may want to put this in a block (```) to avoid weird formatting. You can check how this looks by running `cargo doc --open`

Reply:
done